node/silkworm/db/access_layer.cpp

```diff
@@ -502,22 +578,62 @@ std::optional<ChainConfig> read_chain_config(mdbx::txn& txn) {
     return ChainConfig::from_json(json);
 }
 
+static Bytes head_header_key() {
+    std::string table_name = db::table::kHeadHeader.name;
+    Bytes key{table_name.begin(), table_name.end()};
+    return key;
+}
+
+void write_head_header_hash(mdbx::txn& txn, const evmc::bytes32& hash) {
+    write_head_header_hash(txn, hash.bytes);
+}
+
 void write_head_header_hash(mdbx::txn& txn, const uint8_t (&hash)[kHashLength]) {
     Cursor target(txn, table::kHeadHeader);
-    mdbx::slice key(db::table::kLastHeaderKey);
-    target.upsert(key, to_slice(hash));
+    Bytes key = head_header_key();
+    auto skey = db::to_slice(key);
+
+    target.upsert(skey, to_slice(hash));
 }
 
 std::optional<evmc::bytes32> read_head_header_hash(mdbx::txn& txn) {
     Cursor src(txn, table::kHeadHeader);
-    mdbx::slice key(db::table::kLastHeaderKey);
-    auto data{src.find(key, /*throw_notfound=*/false)};
+    Bytes key = head_header_key();
+    auto skey = db::to_slice(key);
+    auto data{src.find(skey, /*throw_notfound=*/false)};
     if (!data || data.value.length() != sizeof(evmc::bytes32)) {
         return std::nullopt;
     }
     return to_bytes32(from_slice(data.value));
 }
 
+std::optional<evmc::bytes32> read_canonical_hash(mdbx::txn& txn, BlockNum b) {  // throws db exceptions
+    auto hashes_table = db::open_cursor(txn, db::table::kCanonicalHashes);
+    // accessing this table with only b we will get the hash of the canonical block at height b
+    auto key = db::block_key(b);
+    auto data = hashes_table.find(db::to_slice(key), /*throw_notfound*/ false);
+    if (!data) return std::nullopt;  // not found
+    assert(data.value.length() == kHashLength);
+    return to_bytes32(from_slice(data.value));  // copy
+}
+
+void write_canonical_hash(mdbx::txn& txn, BlockNum b, const evmc::bytes32& hash) {
+    Bytes key = db::block_key(b);
+    auto skey = db::to_slice(key);
+    auto svalue = db::to_slice(hash);
+
+    auto hashes_table = db::open_cursor(txn, db::table::kCanonicalHashes);
+    hashes_table.upsert(skey, svalue);
+    hashes_table.close();
+}
+
+void delete_canonical_hash(mdbx::txn& txn, BlockNum b) {
+    auto hashes_table = db::open_cursor(txn, db::table::kCanonicalHashes);
```

Comment:
Please use here db::Cursor ctor so we can take advantage of cursors pool

Reply:
accepted